flashlight.go

```diff
@@ -437,46 +437,137 @@ func (f *Flashlight) Run(httpProxyAddr, socksProxyAddr string,
+	// Until we know our country, default to IR which has all detection rules
+	log.Debug("Defaulting detour country to IR until real country is known")
+	detour.SetCountry("IR")
+
+	stop, err := f.StartBackgroundServices()
+	if err != nil {
+		log.Errorf("Unable to start background services: %v", err)
+	}
 	defer stop()
 
 	f.RunClientListeners(httpProxyAddr, socksProxyAddr, afterStart, onError)
 }
 
-// Starts background services like config fetching
-func (f *Flashlight) StartBackgroundServices() func() {
-	log.Debug("Starting client proxy background services")
+// StartBackgroundServices starts the goroutine monitoring, bypass, and config fetching background
+// services and returns a function that can be called to stop them.
+func (f *Flashlight) StartBackgroundServices() (func(), error) {
+	log.Debug("Starting background services")
 	// check # of goroutines every minute, print the top 5 stacks with most
 	// goroutines if the # exceeds 800 and is increasing.
 	stopMonitor := goroutines.Monitor(time.Minute, 800, 5)
+	stopGlobalConfigFetch := f.startGlobalConfigFetch()
 
-	stopBypass := bypass.Start(f.addProxyListener, f.configDir, f.userConfig)
-
-	stopConfigFetch := f.startConfigFetch()
-	geolookup.EnablePersistence(filepath.Join(f.configDir, "latestgeoinfo.json"))
-	geolookup.Refresh()
+	stopBypass := services.StartBypassService(f.addProxyListener, f.configDir, f.userConfig)
+	stopConfigService, err := f.startConfigService()
+	if err != nil {
+		return func() {
+			stopMonitor()
+			stopBypass()
+			stopGlobalConfigFetch()
+		}, err
+	}
 
 	return func() {
-		stopConfigFetch()
 		stopMonitor()
 		stopBypass()
+		stopGlobalConfigFetch()
+		stopConfigService()
+	}, nil
+}
+
+func (f *Flashlight) startConfigService() (services.StopFn, error) {
+	obfuscate, _ := f.flagsAsMap["readableconfig"].(bool)
+	handler, err := proxyconfig.Init(f.configDir, obfuscate)
+	if err != nil {
+		return nil, err
 	}
+
+	fn := func(old, new *proxyconfig.ProxyConfig) {
+		var country string
+		if old != nil {
+			country = old.GetCountry()
+		}
+
+		// update the country if it has changed
+		if nc := new.GetCountry(); nc != country {
+			log.Debugf("Setting detour country to %v", nc)
+			detour.SetCountry(nc)
+		}
+
+		proxyMap := f.convertNewProxyConfToOld(new.GetProxy().GetProxies())
+		f.notifyProxyListeners(proxyMap, config.Fetched)
+	}
+	conf, _ := proxyconfig.GetConfig(0)
+	fn(nil, conf)
+
+	proxyconfig.OnConfigChange(fn)
+
+	configOpts := &services.ConfigOptions{
+		OriginURL:    "",
+		UserConfig:   f.userConfig,
+		Sticky:       false,
+		RoundTripper: proxied.ParallelPreferChained(),
+	}
+
+	setConfigFlagOpts(configOpts, f.flagsAsMap)
+	return services.StartConfigService(handler, configOpts)
+}
+
+// setConfigFlagOpts sets the OriginURL, Sticky, and Obfuscate config options based on the input flags.
+func setConfigFlagOpts(opts *services.ConfigOptions, flags map[string]interface{}) {
+	toBool := func(v interface{}) bool {
+		b, _ := v.(bool)
```

Comment:
Yeah, I'll update this to check for nil.

Reply:
Wait, actually, no, this won't panic. Because I'm using the test form (with the optional second return value) it doesn't panic. It sets `b` to false (zero value of the type I'm trying to assert to) and `ok` (which I toss) to false because it failed.

https://go.dev/play/p/R7tPQiqjYA1